[Gutenberg] Add Popular homepage block.

miina opened the pull request:
~~Based on #12.~~
Fixes #23.

Intends to:
- [x] Enable filtering posts by `rating`;
- [x] Pull `adventure` post data dynamically;
- [x] Frontend render for Top adventures;
- [x] Display the correct term;

This PR is blocked by #7.

miina committed:
Start popular posts block.

miina committed:
Merge remote-tracking branch 'origin/feature/12-gutenberg_blocks_base' into feature/23-popular_block

miina committed:
Merge CPT & resolve conflicts.

miina committed:
Add some adventure meta. Pull data dynamically.

miina committed:
Add PHP front-end rendering.

miina committed:
Fix CS.

miina committed:
Remove incorrect import.

miina committed:
Fix pulling correct term. Fix metabox bug.

miina committed:
Merge develop.

kienstra commented on includes/class-amp-travel-cpt.php:
```diff
@@ -0,0 +1,288 @@
+	 * Adds meta boxes for adventure post type.
+	 */
+	public function add_adventure_meta_boxes() {
+		add_meta_box( 'amp_travel_adventure_meta', __( 'Adventure details' ), array( $this, 'adventure_meta_box_html' ), 'adventure', 'side' );
```
Could you please add a text domain:
```php
__( 'Adventure details', 'travel' ),
```

kienstra commented on includes/class-amp-travel-cpt.php:
```diff
@@ -0,0 +1,288 @@
+	 */
+	public function adventure_meta_box_html() {
+		global $post;
+		$adventure_custom = get_post_custom( $post->ID );
```
How about calling [get_post_custom()](https://github.com/WordPress/wordpress-develop/blob/33979450aca7de46e784eb2d422b8772b24ebda3/src/wp-includes/post.php#L2018) without an argument? It looks like that defaults to the global post ID, and it worked locally in briefly testing it.

kienstra commented on includes/class-amp-travel-cpt.php:
```diff
@@ -0,0 +1,288 @@
+		$adventure_custom = get_post_custom( $post->ID );
+		$price            = isset( $adventure_custom['amp_travel_price'][0] ) ? $adventure_custom['amp_travel_price'][0] : '';
+		?>
+		<label for='amp_travel_price'><?php esc_attr_e( 'Price (USD)', 'travel' ); ?></label>
```
Could this be `esc_html_e()` instead of `esc_attr_e()`?

kienstra commented on includes/class-amp-travel-cpt.php:
```diff
@@ -0,0 +1,288 @@
+			}
+
+			if ( isset( $_POST['amp_travel_price'] ) ) {
+				update_post_meta( $post->ID, 'amp_travel_price', esc_attr( $_POST['amp_travel_price'] ) );
```
Could this use `get_the_ID()` instead of `$post->ID`? This works locally.

kienstra commented on includes/class-amp-travel-cpt.php:
```diff
@@ -0,0 +1,288 @@
+	 * Saves the custom meta.
+	 */
+	public function save_adventure_post() {
+		if ( ! empty( $_POST ) ) {
```
This might not be needed. If `$_POST` is empty, the function will return in the next conditional when `isset( $_POST['amp_travel_price'] )` is `false`.

kienstra commented on includes/class-amp-travel-cpt.php:
```diff
@@ -0,0 +1,288 @@
+	 * @return mixed Collection params.
+	 */
+	public function filter_rest_adventure_collection_params( $query_params ) {
+
```
Could you please remove this empty line?

kienstra commented on includes/class-amp-travel-cpt.php:
```diff
@@ -0,0 +1,288 @@
+	 * Filter the REST accepted params to accept ordering by 'rating'.
+	 *
+	 * @param array $query_params Collection params.
+	 * @return mixed Collection params.
```
Wouldn't this return an `array`, just as it's passed an `array` as a parameter?

kienstra commented on includes/class-amp-travel-blocks.php:
```diff
@@ -20,6 +20,115 @@ public function init() {
+		$adventures = get_posts(
+			array(
+				'post_type'   => 'adventure',
+				'numberposts' => 3,
```
Could you store `3` in a variable, as it's used twice?

kienstra commented on includes/class-amp-travel-blocks.php:
```diff
@@ -20,6 +20,115 @@ public function init() {
+
+		if ( ! empty( $attributes['heading'] ) ) {
+			$output .= '<header class="max-width-3 mx-auto px1 md-px2">
+				<h3 class="h1 bold line-height-2 md-hide lg-hide" aria-hidden="true">' . esc_attr( $attributes['heading'] ) . '</h3>
```
Could this and the line below use `esc_html( $attributes['heading'] )` instead of `esc_attr()`?

kienstra commented on includes/class-amp-travel-blocks.php:
```diff
@@ -20,6 +20,115 @@ public function init() {
+								</div>
+							</div>
+							<div class="h2 line-height-2 mb1">
+								<span class="travel-results-result-text">' . esc_attr( get_the_title( $adventure->ID ) ) . '</span>
```
Could this please use `esc_html()`:
```php
esc_html( get_the_title( $adventure->ID ) )
```

kienstra commented on includes/class-amp-travel-blocks.php:
```diff
@@ -20,6 +20,115 @@ public function init() {
+							<div class="h2 line-height-2 mb1">
+								<span class="travel-results-result-text">' . esc_attr( get_the_title( $adventure->ID ) ) . '</span>
+								<span class="travel-results-result-subtext h3">•</span>
+								<span class="travel-results-result-subtext h3">$&nbsp;</span><span class="black bold">' . esc_attr( $price ) . '</span>
```
Could this please use `esc_html()`:
```php
esc_html( $price )
```

kienstra commented on includes/class-amp-travel-blocks.php:
```diff
@@ -20,6 +20,115 @@ public function init() {
+
+			$output .= '</div>
+							</div>
+							<span class="travel-results-result-subtext mr1">' . esc_attr( $comments->approved ) . esc_attr__( ' Reviews' ) . '</span>
```
Could this be something like:
```
sprintf( esc_html__( '%d Reviews', 'travel' ), esc_html( $comments->approved ) )
```

With a translators comment above:
```php
<?php /* translators: %d: The number of reviews */ ?>
```

kienstra commented on includes/class-amp-travel-blocks.php:
```diff
@@ -20,6 +20,115 @@ public function init() {
+							</div>
+							<span class="travel-results-result-subtext mr1">' . esc_attr( $comments->approved ) . esc_attr__( ' Reviews' ) . '</span>
+							<span class="travel-results-result-subtext"><svg class="travel-icon" viewBox="0 0 77 100"><g fill="none" fillRule="evenodd"><path stroke="currentColor" strokeWidth="7.5" d="M38.794 93.248C58.264 67.825 68 49.692 68 38.848 68 22.365 54.57 9 38 9S8 22.364 8 38.85c0 10.842 9.735 28.975 29.206 54.398a1 1 0 0 0 1.588 0z"></path><circle cx="38" cy="39" r="10" fill="currentColor"></circle></g></s ...
+							' . esc_attr( $location ) . '</span>
```
Could you please change this to `esc_html( $location )`

kienstra reviewed (changes requested):
**Good Work, A Few Points**

Hi @miina,
Thanks for adding this 'Popular' homepage block. 

It looks like this PR uses commits from #33, so I tried to keep the scope of the review to the [diff between](https://github.com/xwp/travel/compare/feature/7-declare-cpt...feature/23-popular_block) this PR and #33.

There will probably be more merge conflicts after merging #33. 

I'm trying to merge that first, as it seems like a big dependency.

Thanks!

miina committed:
Merge develop & resolve conflicts.

miina committed:
Fixes from feedback. Remove editable title. Fix style.

miina commented:
@kienstra Thanks for reviewing, made the changes + some other small changes found when testing the block.

miina commented:
Note that the heading of the block is static, looks like the texts of dynamic blocks can't be edited directly as with other blocks (at least with the Gutenberg stable version as of current moment) and making the heading editable is something to look into later.

kienstra committed:
Indent 2 lines, whitespace only.

This was the only remaining issue in the PR.
So I made this change on my own to merge it.

kienstra committed:
Merge in develop, resolve conflicts.

There were conflicts in:
assets/css/editor-blocks.css
includes/class-amp-travel-blocks.php
Retain edits from both develop and this branch:
feature/23-popular_block
Also, change order of register_block_travel_popular()
and render_block_travel_popular()
But make no change in those methods.

kienstra commented:
**Approved**

Hi @miina,
Thanks for your great work here. As you mentioned, the heading of this can't be edited now, as this is a dynamic block. We might revisit this later.

Also, as you mentioned, this depends on Issue #7 (PR #33). That registers the "Adventure" custom post type. But there's no error, this simply doesn't display an adventure.

kienstra committed:
Revert changes to style.css.

These look to have been from PR #33.
And they might be valid.
But defer the discussion of those to that PR.

kienstra commented:
**Deferring Merge Until #33 Approved**

As this has commits from #33, I'm going to try to merge that PR before this.

kienstra committed:
Merge in develop, resolve conflicts.

There were conflics in:
assets/css/editor-blocks.css
includes/class-amp-travel-theme.php
In AMP_Travel_Theme, these were simply from
reorganizing that class in a recent other PR.

kienstra commented:
**Merging**

With #33 merged into `develop`, I merged `develop` into this branch. It's now ready to merge.
